src/Grpc.Net.Client/Internal/Retry/HedgingCall.cs

```diff
@@ -0,0 +1,421 @@
+                        break;
+                    }
+                    else
+                    {
+                        lock (Lock)
+                        {
+                            if (IsRetryThrottlingActive())
+                            {
+                                if (_activeCalls.Count == 0)
+                                {
+                                    CommitCall(CreateStatusCall(GrpcProtocolConstants.ThrottledStatus), CommitReason.Throttled);
+                                }
+                                else
+                                {
+                                    Log.AdditionalCallsBlockedByRetryThrottling(Logger);
+                                }
+                                break;
+                            }
+
+                            // Don't send additional calls if call has been commited.
+                            if (CommitedCallTask.IsCompletedSuccessfully())
+                            {
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                HandleUnexpectedError(ex);
+            }
+            finally
+            {
+                Log.StoppingRetryWorker(Logger);
+            }
+        }
+
+        private async Task HedgingDelayAsync(TimeSpan hedgingDelay)
+        {
+            CompatibilityExtensions.Assert(_hedgingDelayCts != null);
+            CompatibilityExtensions.Assert(_delayInterruptTcs != null);
+
+            while (true)
+            {
+                CompatibilityExtensions.Assert(_hedgingDelayCts != null);
+
+                var completedTask = await Task.WhenAny(Task.Delay(hedgingDelay, _hedgingDelayCts.Token), _delayInterruptTcs.Task).ConfigureAwait(false);
+                if (completedTask != _delayInterruptTcs.Task)
+                {
+                    // Task.Delay won. Check CTS to see if it won because of cancellation.
+                    _hedgingDelayCts.Token.ThrowIfCancellationRequested();
+                    return;
+                }
+                else
+                {
+                    // Cancel the Task.Delay that's no longer needed.
+                    // https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/519ef7d231c01116f02bc04354816a735f2a36b6/AsyncGuidance.md#using-a-timeout
+                    _hedgingDelayCts.Cancel();
+                }
+
+                lock (Lock)
+                {
+                    // Need to recreate the delay TCS/CTS as needed for next cycle.
+                    if (_delayInterruptTcs.Task.IsCompleted)
+                    {
+                        _delayInterruptTcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+                    }
+                    if (_hedgingDelayCts.IsCancellationRequested)
```

Comment:
And this looks like it's always canceled here.

Reply:
If there is no interrupt and the overall call wasn't cancelled then this CTS won't be canceled.